File: tests/test_triggers.py
#!/usr/bin/env python3
"""Tests for triggers and automation (Feature 18)."""

import pytest
from par_term_emu_core_rust import (
    Terminal,
    TriggerAction,
    TriggerMatch,
    Trigger,
    CoprocessConfig,
)


def test_trigger_crud():
    """Test add, list, get, remove triggers."""
    term = Terminal(80, 24)

    # Add trigger
    action = TriggerAction("highlight", {"bg_r": "255", "bg_g": "0", "bg_b": "0"})
    trigger_id = term.add_trigger("errors", r"ERROR", [action])
    assert isinstance(trigger_id, int)
    assert trigger_id > 0

    # List triggers
    triggers = term.list_triggers()
    assert len(triggers) == 1
    assert isinstance(triggers[0], Trigger)
    assert triggers[0].id == trigger_id
    assert triggers[0].name == "errors"
    assert triggers[0].pattern == "ERROR"
    assert triggers[0].enabled is True

    # Get trigger
    t = term.get_trigger(trigger_id)
    assert t is not None
    assert t.id == trigger_id

    # Get nonexistent
    assert term.get_trigger(99999) is None

    # Remove trigger
    assert term.remove_trigger(trigger_id) is True
    assert term.remove_trigger(trigger_id) is False  # already removed
    assert len(term.list_triggers()) == 0


def test_trigger_match_detection():
    """Test that triggers detect patterns in terminal output."""
    term = Terminal(80, 24)
    term.add_trigger("error", r"ERROR:\s+(\S+)", [])

    term.process_str("prefix ERROR: something_went_wrong\n")
    term.process_trigger_scans()

    matches = term.poll_trigger_matches()
    assert len(matches) == 1
    assert isinstance(matches[0], TriggerMatch)
    assert matches[0].row == 0
    assert "ERROR:" in matches[0].text
    assert len(matches[0].captures) == 2  # group 0 + group 1
    assert matches[0].captures[1] == "something_went_wrong"

    # Subsequent poll should be empty
    assert len(term.poll_trigger_matches()) == 0


def test_trigger_action_types():
    """Test constructing all action types from Python."""
    # Highlight
    a1 = TriggerAction(
        "highlight", {"bg_r": "255", "bg_g": "0", "bg_b": "0", "duration_ms": "5000"}
    )
    assert a1.action_type == "highlight"

    # Notify
    a2 = TriggerAction("notify", {"title": "Alert", "message": "Error: $1"})
    assert a2.action_type == "notify"

    # Mark line
    a3 = TriggerAction("mark_line", {"label": "Important"})
    assert a3.action_type == "mark_line"

    # Set variable
    a4 = TriggerAction("set_variable", {"name": "status", "value": "$1"})
    assert a4.action_type == "set_variable"

    # Run command
    a5 = TriggerAction("run_command", {"command": "echo", "args": "hello,world"})
    assert a5.action_type == "run_command"

    # Play sound
    a6 = TriggerAction("play_sound", {"sound_id": "alert", "volume": "80"})
    assert a6.action_type == "play_sound"

    # Send text
    a7 = TriggerAction("send_text", {"text": "response\n", "delay_ms": "100"})
    assert a7.action_type == "send_text"

    # Stop propagation
    a8 = TriggerAction("stop")
    assert a8.action_type == "stop"

    # All should be usable in add_trigger
    term = Terminal(80, 24)
    trigger_id = term.add_trigger(
        "all_actions", "TRIGGER", [a1, a2, a3, a4, a5, a6, a7, a8]
    )
    assert trigger_id > 0


def test_trigger_highlights():
    """Test highlight overlays from trigger actions."""
    term = Terminal(80, 24)
    action = TriggerAction("highlight", {"bg_r": "255", "bg_g": "0", "bg_b": "0"})
    term.add_trigger("test", "HIGHLIGHT", [action])

    term.process_str("HIGHLIGHT this\n")
    term.process_trigger_scans()

    highlights = term.get_trigger_highlights()
    assert len(highlights) == 1
    row, col_start, col_end, fg, bg = highlights[0]
    assert row == 0
    assert bg == (255, 0, 0)
    assert fg is None

    # Clear
    term.clear_trigger_highlights()
    assert len(term.get_trigger_highlights()) == 0


def test_trigger_enable_disable():
    """Test enabling and disabling triggers."""
    term = Terminal(80, 24)
    trigger_id = term.add_trigger("test", "MATCH", [])

    assert term.set_trigger_enabled(trigger_id, False) is True
    t = term.get_trigger(trigger_id)
    assert t.enabled is False

    assert term.set_trigger_enabled(trigger_id, True) is True
    t = term.get_trigger(trigger_id)
    assert t.enabled is True


def test_trigger_invalid_regex():
    """Test that invalid regex raises ValueError."""
    term = Terminal(80, 24)
    with pytest.raises(ValueError):
        term.add_trigger("bad", "[invalid", [])


def test_trigger_invalid_action():
    """Test that invalid action type raises ValueError."""
    term = Terminal(80, 24)
    bad_action = TriggerAction("nonexistent_action")
    with pytest.raises(ValueError):
        term.add_trigger("test", "MATCH", [bad_action])


def test_trigger_repr():
    """Test repr methods."""
    action = TriggerAction("highlight", {"bg_r": "255"})
    assert "highlight" in repr(action)

    term = Terminal(80, 24)
    tid = term.add_trigger("test_repr", "PATTERN", [])
    t = term.get_trigger(tid)
    assert "test_repr" in repr(t)


def test_coprocess_config():
    """Test CoprocessConfig construction."""
    config = CoprocessConfig("cat")
    assert config.command == "cat"
    assert config.args == []
    assert config.cwd is None
    assert config.env == {}
    assert config.copy_terminal_output is True

    config2 = CoprocessConfig(
        "grep",
        args=["ERROR"],
        cwd="/tmp",
        env={"LC_ALL": "C"},
        copy_terminal_output=False,
    )
    assert config2.command == "grep"
    assert config2.args == ["ERROR"]
    assert config2.cwd == "/tmp"
    assert config2.copy_terminal_output is False
    assert "grep" in repr(config2)

File: tests/test_triggers.rs
// Integration tests for trigger system (Feature 18)
use par_term_emu_core_rust::terminal::trigger::TriggerAction;
use par_term_emu_core_rust::terminal::Terminal;

#[test]
fn test_trigger_add_remove() {
    let mut term = Terminal::new(80, 24);
    let id = term
        .add_trigger("test".into(), "ERROR".into(), vec![])
        .unwrap();
    assert_eq!(term.list_triggers().len(), 1);
    assert!(term.get_trigger(id).is_some());
    assert!(term.remove_trigger(id));
    assert!(term.get_trigger(id).is_none());
    assert_eq!(term.list_triggers().len(), 0);
}

#[test]
fn test_trigger_enable_disable() {
    let mut term = Terminal::new(80, 24);
    let id = term
        .add_trigger("test".into(), "MATCH".into(), vec![])
        .unwrap();

    // Process text with trigger enabled
    term.process(b"MATCH here\n");
    term.process_trigger_scans();
    let matches = term.poll_trigger_matches();
    assert_eq!(matches.len(), 1);

    // Disable trigger
    assert!(term.set_trigger_enabled(id, false));

    term.process(b"MATCH again\n");
    term.process_trigger_scans();
    let matches = term.poll_trigger_matches();
    assert_eq!(matches.len(), 0);

    // Re-enable
    assert!(term.set_trigger_enabled(id, true));
    term.process(b"MATCH once more\n");
    term.process_trigger_scans();
    let matches = term.poll_trigger_matches();
    assert_eq!(matches.len(), 1);
}

#[test]
fn test_trigger_scan_match() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger("error".into(), r"ERROR:\s+(.+)".into(), vec![])
        .unwrap();

    term.process(b"prefix ERROR: something went wrong\n");
    term.process_trigger_scans();

    let matches = term.poll_trigger_matches();
    assert_eq!(matches.len(), 1);
    assert_eq!(matches[0].row, 0);
    assert!(matches[0].text.contains("ERROR:"));
    assert_eq!(matches[0].captures.len(), 2); // group 0 + group 1
}

#[test]
fn test_trigger_capture_groups() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "ip".into(),
        r"(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})".into(),
        vec![],
    )
    .unwrap();

    term.process(b"IP: 192.168.1.100\n");
    term.process_trigger_scans();

    let matches = term.poll_trigger_matches();
    assert_eq!(matches.len(), 1);
    assert_eq!(matches[0].captures.len(), 5);
    assert_eq!(matches[0].captures[0], "192.168.1.100");
    assert_eq!(matches[0].captures[1], "192");
    assert_eq!(matches[0].captures[4], "100");
}

#[test]
fn test_trigger_multi_pattern() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger("error".into(), "ERROR".into(), vec![])
        .unwrap();
    term.add_trigger("warn".into(), "WARN".into(), vec![])
        .unwrap();

    term.process(b"ERROR and WARN on same line\n");
    term.process_trigger_scans();

    let matches = term.poll_trigger_matches();
    assert_eq!(matches.len(), 2);
}

#[test]
fn test_trigger_invalid_regex() {
    let mut term = Terminal::new(80, 24);
    let result = term.add_trigger("bad".into(), "[invalid".into(), vec![]);
    assert!(result.is_err());
}

#[test]
fn test_trigger_event_generation() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger("test".into(), "FOUND".into(), vec![])
        .unwrap();

    term.process(b"FOUND it\n");
    term.process_trigger_scans();

    let events = term.poll_events();
    let trigger_events: Vec<_> = events
        .iter()
        .filter(|e| {
            matches!(
                e,
                par_term_emu_core_rust::terminal::TerminalEvent::TriggerMatched(_)
            )
        })
        .collect();
    assert_eq!(trigger_events.len(), 1);
}

#[test]
fn test_trigger_action_highlight() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "test".into(),
        "HIGHLIGHT".into(),
        vec![TriggerAction::Highlight {
            fg: None,
            bg: Some((255, 0, 0)),
            duration_ms: 0, // permanent
        }],
    )
    .unwrap();

    term.process(b"HIGHLIGHT this\n");
    term.process_trigger_scans();

    let highlights = term.get_trigger_highlights();
    assert_eq!(highlights.len(), 1);
    assert_eq!(highlights[0].row, 0);
    assert_eq!(highlights[0].bg, Some((255, 0, 0)));
    assert_eq!(highlights[0].fg, None);
}

#[test]
fn test_trigger_action_notify() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "test".into(),
        r"ERROR: (\S+)".into(),
        vec![TriggerAction::Notify {
            title: "Error Alert".into(),
            message: "Found: $1".into(),
        }],
    )
    .unwrap();

    term.process(b"ERROR: diskfull\n");
    term.process_trigger_scans();

    let notifications = term.notifications();
    assert_eq!(notifications.len(), 1);
    assert_eq!(notifications[0].title, "Error Alert");
    assert_eq!(notifications[0].message, "Found: diskfull");
}

#[test]
fn test_trigger_action_mark_line() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "test".into(),
        "BOOKMARK".into(),
        vec![TriggerAction::MarkLine {
            label: Some("Auto-bookmark".into()),
            color: None,
        }],
    )
    .unwrap();

    term.process(b"BOOKMARK here\n");
    term.process_trigger_scans();

    let bookmarks = term.get_bookmarks();
    assert_eq!(bookmarks.len(), 1);
    assert_eq!(bookmarks[0].label, "Auto-bookmark");
}

#[test]
fn test_trigger_action_set_variable() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "test".into(),
        r"STATUS: (\w+)".into(),
        vec![TriggerAction::SetVariable {
            name: "last_status".into(),
            value: "$1".into(),
        }],
    )
    .unwrap();

    term.process(b"STATUS: RUNNING\n");
    term.process_trigger_scans();

    let vars = term.session_variables();
    assert_eq!(vars.custom.get("last_status"), Some(&"RUNNING".to_string()));
}

#[test]
fn test_trigger_action_stop_propagation() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "test".into(),
        "STOP".into(),
        vec![
            TriggerAction::MarkLine {
                label: Some("Before stop".into()),
                color: None,
            },
            TriggerAction::StopPropagation,
            TriggerAction::MarkLine {
                label: Some("After stop".into()),
                color: None,
            },
        ],
    )
    .unwrap();

    term.process(b"STOP here\n");
    term.process_trigger_scans();

    // Only the first action before StopPropagation should have executed
    let bookmarks = term.get_bookmarks();
    assert_eq!(bookmarks.len(), 1);
    assert_eq!(bookmarks[0].label, "Before stop");
}

#[test]
fn test_trigger_capture_substitution() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "test".into(),
        r"USER:(\w+) CMD:(\w+)".into(),
        vec![TriggerAction::SetVariable {
            name: "user_$1".into(),
            value: "ran_$2".into(),
        }],
    )
    .unwrap();

    term.process(b"USER:alice CMD:deploy\n");
    term.process_trigger_scans();

    let vars = term.session_variables();
    assert_eq!(
        vars.custom.get("user_alice"),
        Some(&"ran_deploy".to_string())
    );
}

#[test]
fn test_highlight_expiry() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "test".into(),
        "BRIEF".into(),
        vec![TriggerAction::Highlight {
            fg: None,
            bg: Some((0, 255, 0)),
            duration_ms: 1, // 1ms - will expire almost immediately
        }],
    )
    .unwrap();

    term.process(b"BRIEF flash\n");
    term.process_trigger_scans();

    // Wait for expiry
    std::thread::sleep(std::time::Duration::from_millis(10));

    let highlights = term.get_trigger_highlights();
    assert_eq!(highlights.len(), 0, "Expired highlights should be filtered");

    // Also test clear_expired_highlights
    term.clear_expired_highlights();
}

#[test]
fn test_trigger_frontend_actions() {
    let mut term = Terminal::new(80, 24);
    term.add_trigger(
        "test".into(),
        "RUN".into(),
        vec![
            TriggerAction::RunCommand {
                command: "echo".into(),
                args: vec!["hello".into()],
            },
            TriggerAction::PlaySound {
                sound_id: "alert".into(),
                volume: 80,
            },
            TriggerAction::SendText {
                text: "response\n".into(),
                delay_ms: 100,
            },
        ],
    )
    .unwrap();

    term.process(b"RUN now\n");
    term.process_trigger_scans();

    let results = term.poll_action_results();
    assert_eq!(results.len(), 3);
}
